keps/sig-node/4817-resource-claim-device-status/README.md

```diff
@@ -0,0 +1,614 @@
+interface name.
+
+`DeviceInfo` being a slice of arbitrary data allows the DRA Driver to store
+device specific data in different formats. For example, a Network Device being
+configured by via a CNI plugin could get its `DeviceInfo` filled with the CNI
+result for troubleshooting purpose and with a Network result in a modern
+standard format (closer to Pod.Status.PodIPs for example) used by 3rd party
+controllers.
+
+For each device, if required, the DRA Driver processing the device allocation
+can report the status of it in the `Status.DeviceStatuses` of the ResourceClaim
+by using the Kubernetes API.
+
+```golang
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the result of that was.
+type ResourceClaimStatus struct {
+	...
+
+	// DeviceStatuses contains the status of each device allocated for this
+	// claim, as reported by the driver. This can include driver-specific
+	// information. Entries are owned by their respective drivers.
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=devicePoolName
+	// +listMapKey=deviceName
+	DeviceStatuses []AllocatedDeviceStatus `json:"deviceStatuses,omitempty" protobuf:"bytes,4,opt,name=deviceStatuses"`
+}
+
+
+// AllocatedDeviceStatus contains the status of an allocated device, if the
+// driver chooses to report it. This may include driver-specific information.
+type AllocatedDeviceStatus struct {
+	// Request is the name of the request in the claim which caused this
+	// device to be allocated. Multiple devices may have been allocated
+	// per request.
+	//
+	// +required
+	Request string `json:"request" protobuf:"bytes,1,rep,name=request"`
+
+	// Driver specifies the name of the DRA driver whose kubelet
+	// plugin should be invoked to process the allocation once the claim is
+	// needed on a node.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	//
+	// +required
+	Driver string `json:"driver" protobuf:"bytes,2,rep,name=driver"`
+
+	// This name together with the driver name and the device name field
+	// identify which device was allocated (`<driver name>/<pool name>/<device name>`).
+	//
+	// Must not be longer than 253 characters and may contain one or more
+	// DNS sub-domains separated by slashes.
+	//
+	// +required
+	Pool string `json:"pool" protobuf:"bytes,3,rep,name=pool"`
+
+	// Device references one device instance via its name in the driver's
+	// resource pool. It must be a DNS label.
+	//
+	// +required
+	Device string `json:"device" protobuf:"bytes,4,rep,name=device"`
+
+	// Conditions contains the latest observation of the device's state.
+	// If the device has been configured according to the class and claim
+	// config references, the `Ready` condition should be True.
+	//
+	// +optional
+	// +listType=atomic
+	Conditions []metav1.Condition `json:"conditions" protobuf:"bytes,5,rep,name=conditions"`
+
+	// DeviceInfo contains Arbitrary driver-specific data.
```

Comment:
```suggestion
	// DeviceInfo contains arbitrary driver-specific data.
```

Reply:
How can a consumer reliably determine how to parse this content? The existing `driver` field above might not be enough: for example, two different drivers "foo.example.com" and "bar.example.com" might both provide data according to some common specification.

A consumer shouldn't have to keep a database that lists whether a driver uses such a common specification.

Reply:
> How can a consumer reliably determine how to parse this content? The existing `driver` field above might not be enough: for example, two different drivers "foo.example.com" and "bar.example.com" might both provide data according to some common specification.
> 
> 
> 
> A consumer shouldn't have to keep a database that lists whether a driver uses such a common specification.
> 
> 

One option is to require a GVK in here. But another is just to say - you shouldn't *be* consuming opaque data. If there is a common spec, it should be a built in type.

Reply:
I think that was the idea yes as it would be quite difficult to cover all use cases with the specs we would provide in this KEP. Letting the driver reporting arbitrary status could still be useful I guess (for troubleshooting for example)?

Reply:
yes. but no one except the driver should rely on that data